### ISSUE TITLE
Add ability to run failed tests

### ISSUE THREAD

**Issue by deepthigc** [getgauge/gauge#105]

**Comment by haroon-sheikh** [comment]
@deepthigc is it possible to prioritise this? I think it's a very important feature to have. Will add a lot of value when comparing to cucumber for example.

**Comment by mahendrakariya** [comment]
Here's how this feature could be implemented.

  1. Gauge should start adding a `.gauge` folder to the project root.
  2. A file named `failed` will be created in `.gauge` folder. This file should be overwritten after every run.
  3. This file will contain information about the failed specs / scenarios. (Sample format below.)
  4. A new flag `--failed` should be added. To run only the failed tests, users should be able to do `gauge --failed specs`. On running the command
      * Gauge should only run the scenarios that have failed and not the entire spec.
      * In case of table driven specification, Gauge should only run the rows/scenarios combination that have failed.
      * In case of `BeforeSuite` or `AfterSuite` failure, Gauge should run all the specs.
      * In case the `.gauge/failed` file is empty (no failures in the previous run), Gauge should throw appropriate error message.
  5. There won't be any change in Gauge's report generation behavior. After execution, Gauge should generate a new set of reports which will contain results only for whatever is executed (failed specs / scenarios) in this run.
  6. Users should be able to pass parallel execution related flags along with `--failed`.
  7. The flags `--table-rows` and `--tags` should not be allowed in combination with `--failed` flag.


### Structure of `.gauge/failed` file

Each line of this file will contain an entry about failed spec/scenario. The spec file paths should be relative to `PROJECT_ROOT`.

  * If a spec has 3 scenarios and 1st and 3rd one failed, following entries should be added.
```
specs/some_spec.md:0
specs/some_spec.md:2
```
  * If a spec has 3 scenarios and all of them failed, following entries should be added.
```
specs/login/some_spec.md:0
specs/login/some_spec.md:1
specs/login/some_spec.md:2
```
  * In a table driven spec with 3 scenarios, if 1st scenario failed for rows 1, 2, 4 and 2nd scenario for rows 1 to 3, following entries should be added.
```
specs/some_spec.md:0 tr:1
specs/some_spec.md:0 tr:2
specs/some_spec.md:0 tr:4
specs/some_spec.md:1 tr:1
specs/some_spec.md:1 tr:2
specs/some_spec.md:1 tr:3
```

* In case of `BeforeSuite` or `AfterSuite` failure,  following entries should be added.
```
specs/
```

**P.S.** We will wait for some time before starting to implement this so as to gather feedback from the users.

**Comment by haroon-sheikh** [comment]
Thanks @mahendrakariya for prioritising this.

I think What would be useful if we could set a flags in env/default.properties `run_failed_tests` (default to false) and `retry_count` (optional) defaults to just 1.

So that when running `gauge specs/` it rerun any failed scenario automatically rather than you having to run it again with `gauge --failed specs`. Although, that should also be possible.

Let me know if something is unclear.

**Comment by mahendrakariya** [comment]
@haroon-sheikh Thanks for your input! We really appreciate it.

However, we feel that this is a very specific use case and that it will break the intuitiveness that comes with running `gauge specs`. We feel that the behavior of `gauge specs` command should not be overloaded by setting some parameters.

But in case you would want to overload the behavior, you should be able to do so by wrapping the command in some (shell / batch) script.

**Comment by haroon-sheikh** [comment]
@mahendrakariya I was just thinking about from Continuous Integration point of view.

Although the second step manually be automated. However, we do not want CI build failing just because some test has failed due to flakiness in the first run.

Hence, it would really be handy to have an optional flag to enable/disable retrying of scenarios.

**Comment by sriv** [comment]
@haroon-sheikh - I understand where you are coming from, but it seems to me that this would just disguise flaky tests (and in turn could yield false positives/negatives). I disagree with this approach, but you are the best judge of your setup, hence I am sure you understand the pros/cons much better than I do in this scenario.

From Gauge's point of view, this would mean that `gauge specs` command could be indeterministic i.e. the behaviour of `gauge specs` would become contextual, which I do not think is a good thing.

Here is my proposal:

- let Gauge support running failed tests explicitly (via @mahendrakariya's proposal [above](https://github.com/getgauge/gauge/issues/105#issuecomment-198221378)).
- You (or anyone else who needs it to automatically run failed tests) could have a wrapper script to do so
- In your CI/CD pipeline, you could re-trigger stages with appropriate options to re-run failed tests. Ex. Go.cd has a feature to [trigger with options](https://docs.go.cd/15.3.0/advanced_usage/trigger_with_options.html) that lets you pass special flags.

Hope this makes sense.

**Comment by haroon-sheikh** [comment]
Thanks @sriv for your input on this.

I completely understand your concern about disguising the bad/flaky tests. As I mentioned earlier, CI (Jenkins) would fail the as soon as first run has some failures. And retrying separately would mean that, we have lost the reports from the first run, and only the failed test results are now published.

I think we should really have an option to run the tests and completely optional rerun of failed scenarios defined within (env.properties) as well having the ability run it explicitly.

I really believe that it would really be useful not just for me but for all gauge'ers. Is there a mailing list or something of users, that we can ask their opinion on this?

**Comment by sriv** [comment]
The mailing list for users exists [here](https://groups.google.com/forum/#!forum/getgauge). You are welcome to start a thread about this idea there.

> And retrying separately would mean that, we have lost the reports from the first run, and only the failed test results are now published.

Ok, I understand your concern a bit more now. But irrespective of how we build this feature, every new run will create a new report. There is no way of updating an existing report.

The only way you could preserve a report in the same physical location is by setting `overwrite_reports=false` in your `env/default.properties`. 

In summary, this feature will save you the effort of figuring out which tests failed, but it will generate a new report, with just the specs/scenarios that have been included in the run.

**Comment by sriv** [comment]
Depends on #356

**Comment by kashishm** [comment]
Running failed scenarios against failed table rows will be done in #416.

**Comment by kashishm** [comment]
Available in nightly > 26/05/2016

**Comment by BugDiver** [comment]
* Failure can be specific to contexts, like env, tags, etc.
* Moreover, in case of tagged execution, some scenarios may get filtered. So the scenario index won't correctly reflect which scenario failed.

A way to mitigate it would be to create a json file, which can contain the execution flags as well as failed items. The failed items will be indicated with its file name and line number.

**Example 1:** Only one scenario failed
```
{
	"Flags": {
		"-env": "chrome"
	},
	"FailedItems": [
		"specs/example.spec:21"
	]
}
```

**Example 2:** If a spec has 3 scenarios and all of them failed, following entries will be added.
```
{
	"Flags": {
                "-env": "chrome",
                "-tags": "tag1&tag3",
        },
	"FailedItems": [
		"some_specs/example.spec:3"
                "some_specs/example.spec:12"
                "some_specs/example.spec:14"
	]
}
```


**Example 3:** In case of `BeforeSuite` or `AfterSuite` failure,  following entries will be added.
```
{
	"Flags": {
		"-parallel": "true"
	},
	"FailedItems": [
		"specs"
	]
}
```

**Comment by sguptatw** [comment]
The html report generates entry in the spec list for each scenario failed in the spec.

**Comment by BugDiver** [comment]
Available in nightly > 03/06/2016
